### ISSUE TITLE
JSON over raw sockets (TCP) cannot parse directory paths with embedded }

### ISSUE THREAD

**Issue by MilhouseVH** [xbmc/xbmc#15562]
### Describe the bug

JSON-RPC fails to parse a request that is parsed by the Webserver.

Take the following JSON-API request

```
{"jsonrpc": "2.0", "params": {"directory": "/etc/abc}", "media": "files", "properties": ["file", "lastmodified"]}, "method": "Files.GetDirectory", "id": "libDirectory"}
```

Note that the value for the `directory` property includes an embedded `}`, ie. `/etc/abc}`.

Post this request to the webserver on port 8080, and it will be parsed correctly (the directory isn't accessible, hence the `"Invalid params."` error) and the response includes the original request `"id"`:
```
{"error":{"code":-32602,"message":"Invalid params."},"id":"libDirectory","jsonrpc":"2.0"}
```

But when sending the same request to JSON-RPC on port 9090, the request fails to parse:
```
{"error":{"code":-32700,"message":"Parse error."},"id":null,"jsonrpc":"2.0"}
```

If the directory path is changed from `/etc/abc}` to `/etc/abc` then JSON-RPC will parse it correctly and return the same `"Invalid params."` response as the webserver.

The above is just a trivial example to demonstrate the parsing difference between the Webserver and JSON-RPC. This has been reported in the forum with a real-life example: https://forum.kodi.tv/showthread.php?tid=158373&pid=2822648#pid2822648

## Expected Behavior

JSON-RPC should be able to parse a request when the directory includes an embedded `}`, just as the webserver is able to, because this is a well formed and valid JSON-API request.

## Actual Behavior

JSON-RPC cannot parse a request when the `directory` property (perhaps any property?) includes an embedded `}`.

## Possible Fix

Clearly there is a difference between the JSON-RPC parsing and Webserver parsing, which should be addressed. Maybe a historical issue has been fixed in the Webserver but not JSON-RPC, but both routes should parse the same, but they don't.

### To Reproduce

This is using `telnet` to communicate with port 9090, it shows the different responses when the `}` is added to the `directory` value:

![s1](https://i.imgur.com/Qf6m9Dk.png)

The debug log shows the following for the initial connection, the valid request and finally the "invalid" request:

```
2019-02-20 17:32:59.829 T:1631474544   DEBUG: JSONRPC Server: New connection detected
2019-02-20 17:32:59.829 T:1631474544    INFO: JSONRPC Server: New connection added
2019-02-20 17:33:03.222 T:1631474544   DEBUG: CUtil::GetMatchingSource: no matching source found for [/etc/abc]
2019-02-20 17:33:12.066 T:1631474544   DEBUG: Previous line repeats 5 times.
2019-02-20 17:33:12.066 T:1631474544   ERROR: JSONRPC: Failed to parse '{"jsonrpc": "2.0", "params": {"directory": "/etc/abc}", "media": "files", "properties": ["file", "lastmodified"]}'
```

### Debuglog

The debuglog can be found here: http://ix.io/1BBj

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [X] Linux
 - [ ] OSX
 - [X] Raspberry-Pi
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name:
 - Kodi version: 18.2-rc1

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by pkerling** [comment]
JSONRPC TCP server is trying to find jsonrpc requests in the input data stream by looking for pairs of `{` and `}`, which is bound to fail. See https://github.com/xbmc/xbmc/blob/master/xbmc/network/TCPServer.cpp#L549

Correct solution: Accumulate data until a valid JSON object is formed.

**Comment by MilhouseVH** [comment]
It's even better when the property value is `/etc/abc{` - now there is no response as presumably the TCP server is waiting for more data.

**Comment by DaveTBlake** [comment]
I find the use of terminology on this issue very confusing.  
>JSON-RPC fails to parse a request that is parsed by the Webserver.

The Webserver does not parse anything, all JSON-RPC requests made to any port are parsed using rapidjson. All "JSON" requests made to Kodi are JSON-RPC requests.

The problem here is **TCP server parsing** as it attempts to form separate JSON-RPC requests from those potentially batched on the input stream. JSON-RPC requests that are successful via webserver fail when the same JSON-RPC request is made via TCP or websocket.

The TCP server needs to find individual JSON-RPC requests. Looking for pairs of `{` `}` or `[` `]` is a reasonable approach, but also needs to correctly handle pairs of quotes `"` allowing for escaped quotes `\"`, and ignore the string contents.  It is the pairs of quotes and escaped quote handling that is currently missing,

**Comment by pkerling** [comment]
> The TCP server needs to find individual JSON-RPC requests. Looking for pairs of `{` `}` or `[` `]` is a reasonable approach, but also needs to correctly handle pairs of quotes `"` allowing for escaped quotes `\"`, and ignore the string contents. It is the pairs of quotes and escaped quote handling that is currently missing,

By this you are effectively reimplementing part of JSON parsing. I'd personally rather leave that to rapidjson to avoid problems like this in the future.

**Comment by DaveTBlake** [comment]
>By this you are effectively reimplementing part of JSON parsing. I'd personally rather leave that to rapidjson to avoid problems like this in the future.

So you just pass the TCP stream to rapidjson and it splits it into separate JSON requests. I don't see how that will work. Sometimes partial JSON can look complete hence rapidjson needs to be passed complete requests, and  individually not in a `[ ... ]` batch.

**Comment by pkerling** [comment]
> So you just pass the TCP stream to rapidjson and it splits it into separate JSON requests. I don't see how that will work.

No. You gather and try to parse until rapidjson returns something. Main concern is whether rapidjson would be fine with having extra data after one full JSON request. And is issuing multiple requests (separately, not as batch) without waiting for a response even allowed with JSONRPC?

> Sometimes partial JSON can look complete hence rapidjson needs to be passed complete requests, and individually not in a `[ ... ]` batch.

Batches should not be a problem. They are also valid JSON.

**Comment by pkerling** [comment]
Side-note: I won't stop you or anyone from implementing and merging the fix as you describe. It's clear that it would be strictly an improvement over the current situation. I just hope to propose a solution that is more robust. I'm afraid I won't have the time to actually implement it though so take it with a grain of salt :)

**Comment by MilhouseVH** [comment]
> I find the use of terminology on this issue very confusing.

Yes sorry about that, I'm not entirely sure why but I've just become accustomed to thinking about the Kodi JSON API in terms of the "RPC port" (9090) and "HTTP" port (8080), which I realise is incorrect and by RPC I mean TCP raw sockets so apologies for the confusion.

However it looks like @pkerling has identified the issue in the TCP Server, although I'm not sure what - if anything - should be done about it as it's not a huge problem (it's been there for years!) but it will be (and is) an issue for a small number of users, and the current implementation can be improved to avoid the current weakness (if anyone wants to have a go!)

**Comment by DaveTBlake** [comment]
>an issue for a small number of users, and the current implementation can be improved to avoid the current weakness (if anyone wants to have a go!)

So I did have a go. :) 
Please test #15591 
But @MilhouseVH could you possibly edit the title and desc of this issue to say TCP not JSON-RPC (or may I do it), it is kind of niggling me, JSON is innocent!

**Comment by MilhouseVH** [comment]
> But @MilhouseVH could you possibly edit the title
Done

**Comment by pkerling** [comment]
> Yes sorry about that, I'm not entirely sure why but I've just become accustomed to thinking about the Kodi JSON API in terms of the "RPC port" (9090) and "HTTP" port (8080), which I realise is incorrect and by RPC I mean TCP raw sockets so apologies for the confusion.

I'm not sure what problem you guys had with the title, JSONRPC is pretty much defined as "JSON over raw sockets" :confused: When you are using the HTTP port, you are not doing JSONRPC.

**Comment by pkerling** [comment]
> When you are using the HTTP port, you are not doing JSONRPC.

I take that back. JSONRPC is transport-agnostic. Sorry.

**Comment by MilhouseVH** [comment]
Fixed by #15591.
